File: py/iai.sh.cn/721.cpp
/**
 * 链接：https://iai.sh.cn/problem/721
 */
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
const LL INF = 1e18;
const LL N = 3e5 + 5;

int n, q, _clock = 1;
vector<pair<LL, LL>> g[N], query[N];
LL parent[N], fa[N], dis[N], vis[N], ans[N], cnt_q = 0;

LL _find(LL x, LL* p) {
  if (p[x] == x) return x;
  return p[x] = _find(p[x], p);
}

void dfs(LL root, LL pre, LL v) {
  dis[root] = v;
  for (LL i = 0; i < g[root].size(); i++) {
    LL nx = g[root][i].first, c = g[root][i].second;
    if (nx == pre) continue;
    dfs(nx, root, v + c);
  }
}

void tarjan(LL u) {
  for (LL i = 0; i < g[u].size(); i++) {
    LL nx = g[u][i].first;
    if (vis[nx] == _clock) continue;
    vis[nx] = _clock;
    tarjan(nx);
    fa[_find(nx, fa)] = _find(u, fa);
  }
  for (LL i = 0; i < query[u].size(); i++) {
    LL e = query[u][i].first, idx = query[u][i].second;
    if (vis[e] == _clock) {
      LL lca = _find(e, fa);
      ans[idx] = dis[u] + dis[e] - 2 * dis[lca];
    }
  }
}

int main() {
  cin >> n >> q;
  for (int i = 1; i <= n; i++) {
    parent[i] = i;
    fa[i] = i;
  }

  while (q--) {
    LL x, y, w;
    cin >> x >> y;
    if (getchar() == ' ') {
      cin >> w;
      LL r1 = _find(x, parent), r2 = _find(y, parent);
      parent[r1] = r2;
      g[x].push_back(make_pair(y, w));
      g[y].push_back(make_pair(x, w));
    } else {
      if (_find(x, parent) != _find(y, parent)) {
        ans[cnt_q] = -1;
      } else {
        query[x].push_back(make_pair(y, cnt_q));
        query[y].push_back(make_pair(x, cnt_q));
      }
      ++cnt_q;
    }
  }
  for (int i = 1; i <= n; i++) {
    if (_find(i, parent) == i) {
      g[0].push_back(make_pair(i, 0));
      g[i].push_back(make_pair(0, 0));
    }
  }
  dfs(0, 0, 0);

  tarjan(0);
  for (int i = 0; i < cnt_q; i++) {
    cout << ans[i] << endl;
  }
  return 0;
}

File: py/iai.sh.cn/721.py
n, q = map(int, input().split())
parent = [i for i in range(n + 1)]
g = [[] for _ in range(n + 1)]
ans = []


def find(x, parent):
    if parent[x] == x: return x
    parent[x] = find(parent[x], parent)
    return parent[x]


query = [[] for _ in range(n + 1)]
for i in range(q):
    l = list(map(int, input().split()))
    if len(l) == 2:
        x, y = l
        r1, r2 = find(x, parent), find(y, parent)
        if r1 == r2:
            query[x].append([y, len(ans)])
            query[y].append([x, len(ans)])
        ans.append(-1)
    else:
        x, y, w = l
        r1, r2 = find(x, parent), find(y, parent)
        parent[r1] = r2
        g[x].append([y, w])
        g[y].append([x, w])

vis = [False] * (n + 1)
for i in range(1, n + 1):
    if not vis[find(i, parent)]:
        g[0].append([i, 0])
        g[i].append([0, 0])
        vis[find(i, parent)] = True

vis = [False] * (n + 1)
dis = [0] * (n + 1)


def dfs(root, v):
    dis[root] = v
    for nx, c in g[root]:
        if vis[nx]: continue
        vis[nx] = True
        dfs(nx, v + c)


dfs(0, 0)

fa = [i for i in range(n + 1)]
vis = [False] * (n + 1)


def tarjan(u):
    for nx, c in g[u]:  # 遍历与U相邻的所有节点v
        if vis[nx]: continue
        vis[nx] = True
        tarjan(nx)  # 递归v
        fa[find(nx, fa)] = find(u, fa)  # 把v合并到u上，即将v的父节点设为u
    for e, idx in query[u]:  # 遍历与u有询问关系的节点v
        if vis[e]:
            lca = find(e, fa)
            ans[idx] = dis[u] + dis[e] - 2 * dis[lca]
            # print(idx, u, e, lca, dis, dis[u], dis[e], dis[lca])


tarjan(0)
print(*ans, sep='\n')
